### PR TITLE
Detect changes in GitHub repos defined in parts

### PR DESCRIPTION
**Pull request by jkfran**
# Done

Detect changes in GitHub repos defined in parts, https://github.com/canonical-web-and-design/snap-squad/issues/1286

# QA

Take your time to QA this.

- Check out this branch
- Create .env.local file and ask me for the content (You need new values for this PR)
- Run: `docker build . --tag test`
- Run: `docker run --env-file .env.local test` this will take a while
- It will output everything the script is doing
- Manually check that what is doing is right

### CODE REVIEW

File: poller.py [hunk 1]
```diff
@@ -1,37 +1,115 @@
...
+        return False
+
+    snap_name = snap["store_name"]
+
+    if not snap["store_upload"]:
+        logging.info(f"{snap_name}: Can't be publish from Launchpad")
+        return False
+
+    if not github.is_github_repository_url(snap["git_repository_url"]):
+        logging.info(f"{snap_name}: Is not ussing GitHub")
```
Comment (Lukewh) [inline poller.py:42]:
```suggestion
        logging.info(f"{snap_name}: Is not using GitHub")
```

File: src/github.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        """
+        if (
+            isinstance(url, str)
+            and url.startswith("https://github.com")
+            and url.count("/") == 4
+        ):
+            return True
+        return False
+
+    def get_default_branch(self, owner, repo):
```
Comment (Lukewh) [inline src/github.py:61]: nice

**Review by Lukewh — APPROVED** [review]
:+1: one typo but otherwise looks good.

A couple of thoughts for the future:
- Could we prefix log messages with something that can be grep'd like "SKIP", "BUILD", "NO_NAME" or something?
- Could getting the info from github be more concise using graphql? Would it help reduce the number of calls?

**Comment by jkfran** [comment]
@Lukewh 
>     * Could we prefix log messages with something that can be grep'd like "SKIP", "BUILD", "NO_NAME" or something?
I think this is important and I will create an issue for this.

>     * Could getting the info from github be more concise using graphql? Would it help reduce the number of calls?
After experimenting with this in Snapcraft, I think we wouldn't have much of an advantage. Probably everything here could be in GraphQL, in Snapcraft I found that some queries are not yet available (WebHooks) and we had to switch one big query because it was not as fast as a few REST API queries, but it is defenetly something to keep in mind.

**Comment by Lukewh** [comment]
> After experimenting with this in Snapcraft, I think we wouldn't have much of an advantage. Probably everything here could be in GraphQL, in Snapcraft I found that some queries are not yet available (WebHooks) and we had to switch one big query because it was not as fast as a few REST API queries, but it is defenetly something to keep in mind.

I'm less bothered by the speed (this is a cron, nobody is relying on it being quick), I'm more bothered about the number of requests and hitting the API limit - would GraphQL help with that?

**Pull request merged**